### PR TITLE
RAM Refactoring Issue #541

### PR DESCRIPTION
**Pull request by b-scholz**

### CODE REVIEW

**Comment by b-scholz** [comment]
Index calculations for relations are performed by an RamAnalysis such that interpreter and synthesiser can use them independently. The index calculations are based on a reduction to Dilworth's minimum chain cover problem outlined here:
https://arxiv.org/abs/1709.03685

File: src/IndexSetAnalysis.cpp [hunk 1]
```diff
@@ -0,0 +1,348 @@
...
+
+            // add new order
+            orders.push_back(order);
+
+            // register pseudo chain
+            chainToOrder.push_back(Chain());
+            chainToOrder.back().insert(cur);
+        }
+
+        //		std::cout << "Orders: " << orders << "\n";
```
Comment (mmcgr) [inline src/IndexSetAnalysis.cpp]: Delete these comments

File: src/IndexSetAnalysis.h [hunk 2]
```diff
@@ -8,31 +8,98 @@
...
+#include "RamRelation.h"
 #include "RamTypes.h"
+#include "Util.h"
 
+#include <cstring>
+#include <functional>
 #include <iostream>
 #include <limits>
+#include <map>
+#include <queue>
```
Comment (mmcgr) [inline src/IndexSetAnalysis.h]: std::queue isn't used in this header file, only in the cpp. Move the include to the cp

File: src/IndexSetAnalysis.h [hunk 3]
```diff
@@ -8,31 +8,98 @@
...
 
 #pragma once
 
-#include "MaxMatching.h"
+#include "RamAnalysis.h"
+#include "RamRelation.h"
 #include "RamTypes.h"
+#include "Util.h"
 
+#include <cstring>
```
Comment (mmcgr) [inline src/IndexSetAnalysis.h]: fairly sure cstring, limits, and sstream aren't used and can be removed

File: src/IndexSetAnalysis.h [hunk 4]
```diff
@@ -171,4 +241,35 @@ class AutoIndex {
...
+    static constexpr const char* name = "index-analysis";
+
+    /** run analysis */
+    void run(const RamTranslationUnit& translationUnit) override;
+
+    /** print analysis */
+    void print(std::ostream& os) const override;
+
+    /** get indexes */
+    IndexSet& getIndexes(const RamRelation& rel) {
```
Comment (mmcgr) [inline src/IndexSetAnalysis.h]: Can this be pushed into the cpp?

File: src/Makefile.am [hunk 5]
```diff
@@ -129,6 +127,7 @@ souffle_sources = \
               InterpreterInterface.h                    \
               InterpreterIndex.h                        \
               InterpreterRecords.cpp InterpreterRecords.h \
+			  IndexSetAnalysis.cpp   IndexSetAnalysis.h \
```
Comment (mmcgr) [inline src/Makefile.am]: convert spaces to tabs

File: src/MaxMatching.cpp [hunk 6]
```diff
@@ -15,69 +15,4 @@
...
-        }
-
-        distance[u] = INF;
-        return false;
-    }
-    return true;
-}
-
-}  // end of namespace souffle
+namespace souffle {}  // end of namespace souffle
```
Comment (mmcgr) [inline src/MaxMatching.cpp]: Is this file still needed?

File: src/Synthesiser.cpp [hunk 7]
```diff
@@ -125,7 +125,7 @@ bool useNoIndex() {
     static bool flag = std::getenv(ENV_NO_INDEX);
     static bool first = true;
     if (first && flag) {
-        std::cout << "WARNING: indices are ignored!\n";
+        std::cout << "WARNING: indexes are ignored!\n";
```
Comment (mmcgr) [inline src/Synthesiser.cpp]: indices is the more common usage, outside of the US, and generally preferred in math and technical contexts. Both are acceptable, but why change?

File: src/Synthesiser.cpp [hunk 8]
```diff
@@ -141,33 +141,7 @@ static const std::string getOpContextName(const RamRelation& rel) {
...
-        return data.begin();
-    }
-
-    iterator end() {
-        return data.end();
-    }
-};
-
-std::string getRelationType(const RamRelation& rel, std::size_t arity, const AutoIndex& indices) {
+std::string getRelationType(const RamRelation& rel, std::size_t arity, const IndexSet& indexes) {
```
Comment (mmcgr) [inline src/Synthesiser.cpp]: As above, except in code we use indices even more. Currently it's at 0 uses of 'indexes' and many of 'indices'.

File: src/MaxMatching.cpp (hunk 6 above)
Comment (b-scholz) [reply src/MaxMatching.cpp]: No - I will remove it.

File: src/IndexSetAnalysis.cpp (hunk 1 above)
Comment (b-scholz) [reply src/IndexSetAnalysis.cpp]: fixed.

File: src/IndexSetAnalysis.h (hunk 2 above)
Comment (b-scholz) [reply src/IndexSetAnalysis.h]: I have removed even more unnecessary includes.

File: src/IndexSetAnalysis.h (hunk 3 above)
Comment (b-scholz) [reply src/IndexSetAnalysis.h]: removed.

File: src/Makefile.am (hunk 5 above)
Comment (b-scholz) [reply src/Makefile.am]: removed.

File: src/Synthesiser.cpp (hunk 7 above)
Comment (b-scholz) [reply src/Synthesiser.cpp]: I would prefer US spelling - it is the dominant version of English.

File: src/Synthesiser.cpp (hunk 8 above)
Comment (b-scholz) [reply src/Synthesiser.cpp]: I agree - we need to be consistent - either we use English or American spelling in our source code. Perhaps we start a voting process for this.
